File: 0125-valid-palindrome/0125-valid-palindrome.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        int l = 0; 
        int r = s.length() - 1;

        while (l < r) {
            // If left character is not alphanumeric, skip it and restart loop
            if (!isalnum(s[l])) {
                l++;
                continue; 
            }
            
            // If right character is not alphanumeric, skip it and restart loop
            if (!isalnum(s[r])) {
                r--;
                continue;
            }

            // Now both s[l] and s[r] are guaranteed to be valid alphanumeric characters
            if (tolower(s[l]) == tolower(s[r])) {
                l++; 
                r--;
            } else {
                return false;
            }
        }

        return true;
    }
};

File: 0125-valid-palindrome/0125-valid-palindrome.py
class Solution:
    # 1. Valid Palindrome (LeetCode 125) using Two Pointers
    def isPalindrome(self, s: str) -> bool:
        # Initialize two pointers: 'left' at the beginning and 'right' at the end of the string
        left = 0
        right = len(s) - 1
        
        # Loop until the two pointers meet in the middle
        while left < right:
            # Move the left pointer forward if the character is not alphanumeric
            if not s[left].isalnum():
                left += 1
            # Move the right pointer backward if the character is not alphanumeric
            elif not s[right].isalnum():
                right -= 1
            # Both characters are alphanumeric; compare them case-insensitively
            else:
                if s[left].lower() != s[right].lower():
                    # Return False if characters do not match
                    return False
                # Move both pointers inward if they match
                left += 1
                right -= 1
                
        # Return True if the entire string has been checked successfully
        return True